crates/pallet-grandpa-finality-verifier/src/grandpa.rs

```diff
@@ -111,3 +113,174 @@ pub struct AuthoritySet {
+    finalized_target: (Header::Hash, Header::Number),
+    authorities_set_id: SetId,
+    authorities_set: &VoterSet<AuthorityId>,
+    justification: &GrandpaJustification<Header>,
+) -> Result<(), Error>
+where
+    Header::Number: finality_grandpa::BlockNumberOps,
+{
+    // ensure that it is justification for the expected header
+    if (
+        justification.commit.target_hash,
+        justification.commit.target_number,
+    ) != finalized_target
+    {
+        return Err(Error::InvalidJustificationTarget);
+    }
+
+    let mut chain = AncestryChain::new(&justification.votes_ancestries);
+    let mut signature_buffer = Vec::new();
+    let mut votes = BTreeSet::new();
+    let mut cumulative_weight = 0u64;
+    for signed in &justification.commit.precommits {
+        // authority must be in the set
+        let authority_info = match authorities_set.get(&signed.id) {
+            Some(authority_info) => authority_info,
+            None => {
+                // just ignore precommit from unknown authority as
+                // `finality_grandpa::import_precommit` does
+                continue;
+            }
+        };
+
+        // check if authority has already voted in the same round.
+        //
+        // there's a lot of code in `validate_commit` and `import_precommit` functions inside
+        // `finality-grandpa` crate (mostly related to reporting equivocations). But the only thing
+        // that we care about is that only first vote from the authority is accepted
+        if !votes.insert(signed.id.clone()) {
+            continue;
+        }
+
+        // everything below this line can't just `continue`, because state is already altered
+
+        // precommits aren't allowed for block lower than the target
+        if signed.precommit.target_number < justification.commit.target_number {
+            return Err(Error::PrecommitIsNotCommitDescendant);
+        }
+        // all precommits must be descendants of target block
+        chain = chain.ensure_descendant(
+            &justification.commit.target_hash,
+            &signed.precommit.target_hash,
+        )?;
+        // since we know now that the precommit target is the descendant of the justification
+        // target, we may increase 'weight' of the justification target
+        //
+        // there's a lot of code in the `VoteGraph::insert` method inside `finality-grandpa` crate,
+        // but in the end it is only used to find GHOST, which we don't care about. The only thing
+        // that we care about is that the justification target has enough weight
+        cumulative_weight = cumulative_weight.checked_add(authority_info.weight().0.into()).expect(
+            "sum of weights of ALL authorities is expected not to overflow - this is guaranteed by\
+				existence of VoterSet;\
+				the order of loop conditions guarantees that we can account vote from same authority\
+				multiple times;\
+				thus we'll never overflow the u64::MAX;\
+				qed",
```

Comment:
Is it though if `VoterSet` itself is a user-provided input? I'm not convinced from the comment and quick look at `VoterSet::new`.

Reply:
It is not user provided but rather constructed from the current authorites we will store as part of the state. This sum of upto the previous header is gauranted to not overflow since we do the sum inside the new.

Would look like this - https://github.com/paritytech/parity-bridges-common/blob/master/modules/grandpa/src/lib.rs#L428

Reply:
Hm, true, but as far as this function is concerned in isolation is that statically guaranteed? If no then this is not a panic-safe code.

Reply:
With in the context of this pallet, it should be safe since the inputs are guaranteed to be valid but I see your concern in case this is exposed outside. I'll keep this as is in this PR and will adjust this in the next one since that would pretty much show the use case for this function

Reply:
Should be fixed here: https://github.com/subspace/subspace/commit/d65401bde5d6ac7a3f57d5541d0d62a087c38b38

Reply:
When you read code, you usually read a single function. So if one function has assumptions about its use in the other place and not the other way around, bugs will happen sooner or later.

Function should either be able to guarantee itself that nothing will panic or it should return `Result<>`, API should be as misuse resistant as possible IMHO.